### PR TITLE
Java stubs

### PR DESCRIPTION
**Pull request by mbolotov**
Add ability to automatically generate java-stubs for each contract to easy call solidity functions from java side

### CODE REVIEW

**Comment by SerCeMan** [comment]
Just a general question, what do you think of using https://github.com/web3j/web3j to generate java classes instead of writing code for this? A lot of java projects already use web3j, and it might be beneficial to share the same code. 

Also, I'm sure @conor10 will be happy to answer any questions about the integration.

**Comment by mbolotov** [comment]
Okay, I'll look into it in a few days

**Comment by mbolotov** [comment]
Hey, @SerCeMan 
I've tried web3j's java stuff generation. It looks quite good and comprehensive, but I believe this way does not fit exactly to the one we are striving to - to keep calling contracts simple as much as possible (aren't we? )
With this PR, one can just write 
```Object result = ContractsRepository.getInstance().submitMyContract().myFunction(1);```
And execute this code without any additional configuration.
With web3j, a person has to setup a node to be able to do anything within web3j.

I believe this our approach is quite good for just playing with contracts as well as simple unit testing them.
While the web3j one is more suitable for use in production (or testing them on an integration level or higher)
But I have to say here I have no real experience in contract development :)

What do you think about it?
Do you think Conor may be interested in creating a way to simplify calling contracts?

**Comment by 0v1se** [comment]
Little offtopic, guys: 
besides web3j there is https://github.com/0v1se/scalether 
Basically it does the same thing as web3j, but it also supports non-blocking operations.

For example, we develop apps with spring 5 webflux and use scalether with Mono/Flux processing.
It has code generation too, via standalone app or maven plugin.

It doesn't have any documentation right now, it's planned.
So, if you need non-blocking API for ethereum nodes, let me know - I will try to add some documentation for quick start

**Comment by conor10** [comment]
Hey guys, just seen this.

Happy to see how we can get web3j aligned here. I'm very supportive of @SerCeMan work so far on the Solidity plugin.

It's worth mentioning too that we're building on top of web3j with Microsoft on one of their upcoming Azure cloud offering, so the user base is only going to be increasing.

There's also going to be a Gradle plugin available shortly, and some other community projects that are going to be announced shortly.

**Comment by mbolotov** [comment]
Hey all, 
I've added support for web3j generation style. Your comments are welcome.

**Comment by SerCeMan** [comment]
Hey, @mbolotov!

I'll try to release the new version, because the amount of features in the release is significant already, and then I'll get back to this PR. 

The problem that I'm currently trying to understand is that compilation of solidity files is always enabled. So, if one has a java project that for some reason contains a solidity file which can't be compiled (different Solidity version, etc...) the plugin will become unusable. So, probably, we need some support for facets or something else that will allow for enabling/disabling sol compilation explicitly. 

I'm not sure if this PR is a good place to discuss it, but it seems like everyone is here right now.

**Comment by mbolotov** [comment]
@SerCeMan do you mean an ability to disable solidity compilation per project? Just because it's currently possible to disable the one in global settings.

File: src/main/kotlin/me/serce/solidity/ide/interop/DependencyConfigurator.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  private fun refreshProject(project: Project) {
+    ProgressManager.getInstance().run(object : Task.Backgroundable(project, "Generating java stubs") {
+      override fun run(indicator: ProgressIndicator) {
+        updateDepenedencies(project)
+      }
+    })
+  }
+
+  private fun updateDepenedencies(project: Project) {
```
Comment (xaviarias) [inline src/main/kotlin/me/serce/solidity/ide/interop/DependencyConfigurator.kt]: Seems that here is a typo on this method, should be `updateDependencies`.

**Comment by xaviarias** [comment]
Hi, I have been trying the Java stubs generation but run into some issues that might be worth commenting.

First, I have noticed that it does not does not support Solidity 0.4.21. When compiling the OpenZeppelin `Ownable.sol` I got this error:

```
Solidity compilation failed
            /Users/xavi/IdeaProjects/java-stubs/src/main/solidity/Ownable.sol:38:34: Error: Expected token Semicolon got 'LParen'
            emit OwnershipTransferred(owner, newOwner);
            ^
```

Then, when I change the pragma version and remove `emit` to be compliant with the version:

```
Solidity compilation failed
            /Users/xavi/IdeaProjects/java-stubs/src/main/solidity/Ownable.sol:1:1: Error: Source file requires different compiler version (current compiler is 0.4.19-develop.2018.1.12+commit.c4cbbb05.Darwin.appleclang - note that nightly builds are considered to be strictly less than the released version
            pragma solidity ^0.4.19;
            ^----------------------^
```

I have also tried with version `0.4.19-develop` with the same result.

Other features I missed (or could not find) are to configure the output source folder. By default it created a `src-gen` folder.

Otherwise the code looks very good , I just added a comment on a typo.

**Comment by mbolotov** [comment]
Hey, @xaviarias!
Thanks for your comments!
I added an ability to define folder name for stubs to be generated into. But it just a name, not a full path. Do you think the full path is highly demanded here? It's possible to implement as well but a bit difficult.

As for the compiler version, we could think about providing ability to configure standalone solc compiler.
Currently, IDE extracts it from the EthereumJ bundle, as you might see. 
So there is no smooth way to 'upgrade' the solc version (though you can always repack the bundle with you own one J)

**Pull request merged**